### PR TITLE
AwsAuthRefreshable - aws refreshable token

### LINKED ISSUE #150: Refresh EKS tokens.
The skuber client appears to have no way to refresh EKS tokens - unless I am missing how?
If I am not missing how to refresh tokens we need a way to do this.

The following is from AWS....

"Hello,

We have identified applications running in one or more of your Amazon EKS clusters that are not refreshing service account tokens. Applications making requests to Kubernetes API server with expired tokens will fail. You can resolve the issue by updating your application and its dependencies to use newer versions of Kubernetes client SDK that automatically refreshes the tokens. 

What is the problem?

Kubernetes version 1.21 graduated BoundServiceAccountTokenVolume feature [1] to beta and enabled it by default. This feature improves security of service account tokens by requiring a one hour expiry time, over the previous default of no expiration. This means that applications that do not refetch service account tokens periodically will receive an HTTP 401 unauthorized error response on requests to Kubernetes API server with expired tokens. You can learn more about the BoundServiceAccountToken feature in EKS Kubernetes 1.21 release notes [2].

To enable a smooth migration of applications to the newer time-bound service account tokens, EKS v1.21+ extends the lifetime of service account tokens to 90 days. Applications on EKS v1.21+ clusters that make API server requests with tokens that are older than 90 days will receive an HTTP 401 unauthorized  error response.

How can you resolve the issue?

To make the transition to time bound service account tokens easier, Kubernetes has updated the below official versions of client SDKs to automatically refetch tokens before the one hour expiration:

* Go v0.15.7 and later
* Python v12.0.0 and later
* Java v9.0.0 and later
* Javascript v0.10.3 and later
* Ruby master branch
* Haskell v0.3.0.0

We recommend that you update your application and its dependencies to use one of the above client SDK versions if you are on an older version. 

While not an exhaustive list, the below AWS components have been updated to use the newer Kubernetes client SDKs that automatically refetches the token :

* Amazon VPC CNI: v1.8.0 and later
* CoreDNS: v1.8.4 and later
* AWS Load Balancer Controller: v2.0.0 and later
* kube-proxy: v1.21.2-eksbuild.2 and later 

You can find instructions for upgrading Amazon VPC CNI, CoreDNS, AWS Load Balancer Controller and kube-proxy in EKS documentation at [3], [4], [5] and [6] respectively. Please check EKS documentation [2] for list of components and the versions that have addressed this issue. 

How can I identify service accounts using stale tokens?

As of April 20th 2022, we have identified the below service accounts attached to pods in one or more of your EKS clusters using stale (older than 1 hour) tokens. Service accounts are listed in the format : <eks-cluster-arn>|<namespace:serviceaccount>

...<ommited>"

### PR DESCRIPTION
**Pull request by hagay3**
Fix #150 
TODO
1. try to fetch the expiration date from eks api
2. test on eks cluster
3. create a guide for using AWS auth
4. add code examples

### CODE REVIEW

File: client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    val refreshedToken = RefreshableToken(token, DateTime.now.plus(refreshInterval.toMillis))
+    cachedToken = Some(refreshedToken)
+    refreshedToken
+  }
+
+  override def generateToken: String = {
+    try {
+      val credentialsProvider = new AWSStaticCredentialsProvider(new DefaultAWSCredentialsProviderChain().getCredentials)
+      val signer: Signer = SignerFactory.createSigner(SignerFactory.VERSION_FOUR_SIGNER,
+        new SignerParams("sts", Regions.US_EAST_1.getName)) //it needs to be "us_east_1"
```
Comment (brycechesternewman) [inline client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala:50]: Per AWS calls too STS should be regionalized. ref https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html

File: client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+      val tokenService: AWSSecurityTokenServiceClient = AWSSecurityTokenServiceClientBuilder
+        .standard()
+        .withCredentials(credentialsProvider)
+        .build().asInstanceOf[AWSSecurityTokenServiceClient]
+
+      val callerIdentityRequestDefaultRequest = new DefaultRequest[GetCallerIdentityRequest](new GetCallerIdentityRequest(), "sts")
+      val uri = new URI("https", "sts.amazonaws.com", null, null)
```
Comment (brycechesternewman) [inline client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala]: Per AWS calls too STS should be regionalized. ref https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html. in this case $"sts.$region.amazonaws.com".

Comment (hagay3) [reply client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala]: that's true, but it still works with sts.amazonaws.com
with this approach we don't need to require from the users the region name and also there are few regions that don't have sts deployed.
https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html
eu-south-1
me-south-1
ap-southeast-3
ap-east-1
af-south-1

**Review by hagay3 — COMMENTED** [review]
.

File: client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala (hunk 2 above)
Comment (brycechesternewman) [reply client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala]: We should allow for the region to be specified and if it is not specified then default to the global. We can't rely on the global for our needs.

Comment (hagay3) [reply client/src/main/scala/skuber/api/client/token/AwsAuthRefreshable.scala]: agreed i will fix it

**Pull request merged**